src/ray/core_worker/core_worker.cc

```diff
@@ -630,32 +661,35 @@ Status CoreWorker::ExecuteTask(const TaskSpecification &task_spec,
     return_ids.pop_back();
     task_type = TaskType::ACTOR_TASK;
   }
-  bool direct_call = worker_context_.CurrentActorUseDirectCall();
-  status = task_execution_callback_(
-      task_type, func, task_spec.GetRequiredResources().GetResourceMap(), args,
-      arg_reference_ids, return_ids, direct_call, results);
 
-  SetCurrentTaskId(TaskID::Nil());
-  worker_context_.ResetCurrentTask(task_spec);
+  std::vector<std::shared_ptr<RayObject>> return_objects;
+  status = task_execution_callback_(task_type, func,
+                                    task_spec.GetRequiredResources().GetResourceMap(),
+                                    args, arg_reference_ids, return_ids, &return_objects);
 
-  // TODO(edoakes): this is only used by java.
-  if (results->size() != 0 && !direct_call) {
-    for (size_t i = 0; i < results->size(); i++) {
-      ObjectID id = ObjectID::ForTaskReturn(
-          task_spec.TaskId(), /*index=*/i + 1,
-          /*transport_type=*/static_cast<int>(TaskTransportType::RAYLET));
-      if (!Put(*results->at(i), id).ok()) {
-        // NOTE(hchen): `PlasmaObjectExists` error is already ignored inside
-        // Put`, we treat other error types as fatal here.
-        RAY_LOG(FATAL) << "Task " << task_spec.TaskId() << " failed to put object " << id
-                       << " in store: " << status.message();
-      } else {
-        RAY_LOG(DEBUG) << "Task " << task_spec.TaskId() << " put object " << id
-                       << " in store.";
+  for (size_t i = 0; i < return_objects.size(); i++) {
+    // The object is nullptr if it already existed in the object store.
+    if (!return_objects[i]) {
+      continue;
+    }
+    if (return_objects[i]->GetData()->IsPlasmaBuffer()) {
+      if (!Seal(return_ids[i]).ok()) {
+        RAY_LOG(ERROR) << "Task " << task_spec.TaskId() << " failed to seal object "
+                       << return_ids[i] << " in store: " << status.message();
+      }
+    } else if (!worker_context_.CurrentActorUseDirectCall()) {
+      if (!Put(*return_objects[i], return_ids[i]).ok()) {
+        RAY_LOG(ERROR) << "Task " << task_spec.TaskId() << " failed to seal object "
+                       << return_ids[i] << " in store: " << status.message();
```

Comment:
The above logic should be moved to raylet_transport.cc as the logic is really specific to that transport.

Reply:
also currently this function only populate the returning vector when it's direct actor call, but leave it unpopulated otherwise, which is not consistent. I think it's better to move the logic to r raylet_transport.cc so we always populate the returning vector.

Reply:
@zhijunfu we're going to allow direct actor calls to return plasma objects too, so this will be common logic then.

Currently, direct actor calls cannot return objects larger than 4MB due to gRPC limitations.

Reply:
@ericl I see, thanks. That makes sense.

Could we make sure we always populate the output parameter `return_by_value` as long as this function succeeds? That would be more intuitive to user's expectation. 

e.g. We can move above logic to a utility function, and then change `ExecuteTask` to always populate the output param `return_by_value`, and then the caller of this function `RayletTransport` and `DirectActorTransport` can call into the utility function to handle return results.  That seems more reasonable since the where to put these return objects are really specific to the different transports.

Reply:
Agree with the suggestions. I actually have some of these implemented in a separate PR, so would suggest waiting on that instead of changing it here to avoid conflicts.

Reply:
That sounds good, thanks!